Check usability of inotify before selecting it

okapia opened the pull request:
A libinotify compatibility library exists for BSD systems. If this is installed, configure finds sys/inotify.h, tries to use inotify and consequently the build fails to link. By also checking for inotify_init() this problem can be avoided. AC_CHECK_FUNCS seems to be sufficient – it is returning no on FreeBSD and yes on Linux. I've included the change against the generated configure file in the pull request as this appears to be consistent with the project's practices.

Note that a kqueue compatibility library for Linux also appears to exist so the presence of both sys/event.h and sys/inotify.h could come about on Linux. On Linux with this change, configure would still not bother looking for event.h. I've verified this change on Linux. I'm not aware of other compatibility libraries that might be relevant. Solaris has event ports and no compatibility library exists to my knowledge.

Given that there is native support for kqueue, there is no need to handle linking with libinotify. If you strongly believe that it should be used if installed then let me know and I can try a different approach. I can't see why you would want that when the native support exists.

See #18729.

okapia committed:
Check usability of inotify before selecting it

A libinotify compatibility library exists for BSD systems. If this
was installed, configure would find sys/inotify.h but the build
would fail to link unless the library was added to the link line.
By also checking for inotify_init() this problem can be avoided.

Given that there is native support for kqueue, there is no need
to handle linking with libinotify.

vadz commented:
Thanks, I agree with your argument about not even trying to link with `-linotify`, I didn't think this through.

One tiny improvement (and sorry if I'm nitpicking too much) I see here would be to use a single `AC_LINK_IFELSE()` invocation to test for both `sys/inotify.h` and `inotify_init()`, as this would avoid an extra preprocessor run done by `AC_CHECK_HEADERS` (`AC_CHECK_FUNCS` is already `AC_LINK_IFELSE` in disguise anyhow). This is definitely not a showstopper and we could merge this in its current state, but wx configure already takes quite a long time to run, so ideally I'd like to avoid making it even slower.

okapia commented:
Thanks for reviewing. No problem with the nitpick – I've pushed a replacement using a single `AC_LINK_IFELSE()`.

vadz committed:
Cache the result of inotify check in configure

Wrap AC_LINK_IFELSE() in AC_CACHE_CHECK() to skip it if the result is
already cached.

vadz commented:
Sorry, but I've pushed another small change, that I feel confident enough of doing without testing under FreeBSD -- but please let me know if I was wrong and this broke something.

Otherwise I'll merge this once the CI builds finish. Thanks again!

okapia commented:
I've quickly tested with your additional change under FreeBSD and it all looks good. Thanks
